### PR TITLE
ARROW-6961: [C++][Gandiva] Add string lower function in Gandiva

### PR DESCRIPTION
**Pull request by pprudhvi**

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -168,6 +168,31 @@ char* upper_utf8(int64 context, const char* data, int32 data_len, int32_t* out_l
   return ret;
 }
 
+// Convert a utf8 sequence to lower case.
+// TODO : This handles only ascii characters.
+FORCE_INLINE
+const char* lower_utf8(int64 context, const char* data, int32 data_len,
+                       int32_t* out_len) {
+  char* ret = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, data_len));
+  if (ret == nullptr) {
```
Comment (pitrou) [inline cpp/src/gandiva/precompiled/string_ops.cc]: While we are at it, can you add the same error handling to `upper_utf8` above?

**Review by pitrou — CHANGES_REQUESTED** [review]
Just a single comment.

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 1 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/string_ops.cc]: done. Thanks for reviewing

**Review by pitrou — APPROVED** [review]

File: cpp/src/gandiva/precompiled/string_ops_test.cc [hunk 2]
```diff
@@ -258,4 +258,23 @@ TEST(TestStringOps, TestConcat) {
   EXPECT_EQ(std::string(out_str, out_len), "abcd\na");
   EXPECT_FALSE(ctx.has_error());
 }
+
+TEST(TestStringOps, TestLower) {
+  gandiva::ExecutionContext ctx;
```
Comment (fsaintjacques) [inline cpp/src/gandiva/precompiled/string_ops_test.cc]: This test will raise valgrind memory leak errors.

**Comment by pprudhvi** [comment]
@pitrou some builds are failing. I retriggered a couple of times by closing and reopening, each time different builds are breaking. All builds succeeded across all runs but not in any single run, Could you please merge this?

**Comment by pitrou** [comment]
@pprudhvi I'll do as soon as Travis finishes running on this :-)

**Pull request closed without merging**